Mathlib/Topology/MetricSpace/CoveringNumbers.lean

```diff
@@ -217,6 +228,142 @@ lemma packingNumber_singleton (ε : ℝ≥0) (x : X) : packingNumber ε {x} = 1
+    simp only [hC, iInf_lt_top, encard_lt_top_iff, Subtype.exists, exists_prop]
+    exact ⟨C', ⟨hC'_subset, hC'_cover⟩, hC'_fin⟩
+  · rw [hC]
+    simp_rw [iInf_subtype, iInf_and]
+    rfl
+
+open Classical in
+/-- A finite internal `ε`-cover of a set `A` by closed balls with minimal cardinality.
+It is defined as the empty set if no such finite cover exists. -/
+noncomputable
+def minimalCover (ε : ℝ≥0) (A : Set X) : Set X :=
+  if h : coveringNumber ε A ≠ ⊤ then (exists_set_encard_eq_coveringNumber h).choose else ∅
+
+lemma minimalCover_subset : minimalCover ε A ⊆ A := by
+  by_cases h : coveringNumber ε A ≠ ⊤
+  · simp only [minimalCover, ne_eq, h, not_false_eq_true, ↓reduceDIte]
+    exact (exists_set_encard_eq_coveringNumber h).choose_spec.1
+  · simp [minimalCover, h]
+
+lemma finite_minimalCover :
+    (minimalCover ε A).Finite := by
+  by_cases h : coveringNumber ε A ≠ ⊤
+  · simp only [minimalCover, ne_eq, h, not_false_eq_true, ↓reduceDIte]
+    exact (exists_set_encard_eq_coveringNumber h).choose_spec.2.1
+  · simp [minimalCover, h]
+
+lemma isCover_minimalCover (h : coveringNumber ε A ≠ ⊤) :
+    IsCover ε A (minimalCover ε A) := by
+  simp only [minimalCover, ne_eq, h, not_false_eq_true, ↓reduceDIte]
+  exact (exists_set_encard_eq_coveringNumber h).choose_spec.2.2.1
+
+lemma card_minimalCover (h : coveringNumber ε A ≠ ⊤) :
+    (minimalCover ε A).encard = coveringNumber ε A := by
+  simp only [minimalCover, ne_eq, h, not_false_eq_true, ↓reduceDIte]
+  exact (exists_set_encard_eq_coveringNumber h).choose_spec.2.2.2
+
+end MinimalCover
+
+section MaximalSeparatedSet
+
+lemma exists_set_encard_eq_packingNumber (h : packingNumber ε A ≠ ⊤) :
+    ∃ C, C ⊆ A ∧ C.Finite ∧ IsSeparated ε C ∧ C.encard = packingNumber ε A := by
+  rcases Set.eq_empty_or_nonempty A with hA | hA
+  · simp [hA, packingNumber]
+  have : Nonempty { s : Set X // s ⊆ A ∧ IsSeparated ε s } := by
+    obtain ⟨a, ha⟩ := hA
+    exact ⟨⟨{a}, by simp [ha], by simp⟩⟩
+  let h_exists := ENat.exists_eq_iSup_of_lt_top
+    (f := fun C : { s : Set X // s ⊆ A ∧ IsSeparated ε s } ↦ (C : Set X).encard)
+  simp_rw [packingNumber] at h ⊢
+  simp_rw [iSup_subtype, iSup_and] at h_exists
+  specialize h_exists h.lt_top
+  obtain ⟨C, hC⟩ := h_exists
+  refine ⟨C, C.2.1, ?_, C.2.2, ?_⟩
+  · refine Set.encard_ne_top_iff.mp ?_
+    rwa [hC]
+  · rw [hC]
+
+/-- A finite `ε`-separated subset of a set `A` with maximal cardinality.
+It is defined as the empty set if no such finite subset exists. -/
+noncomputable
+def maximalSeparatedSet (ε : ℝ≥0) (A : Set X) : Set X :=
+  if h : packingNumber ε A ≠ ⊤ then (exists_set_encard_eq_packingNumber h).choose else ∅
+
+lemma maximalSeparatedSet_subset : maximalSeparatedSet ε A ⊆ A := by
+  by_cases h : packingNumber ε A ≠ ⊤
+  · simp only [maximalSeparatedSet, ne_eq, h, not_false_eq_true, ↓reduceDIte]
+    exact (exists_set_encard_eq_packingNumber h).choose_spec.1
+  · simp only [maximalSeparatedSet, h, dite_false, Set.empty_subset]
```

Comment:
This and the next two lemmas can be proven with this, although for it to work on `isSeparated_maximalSeparatedSet`, you have to add
```lean
attribute [grind .] IsSeparated.empty
```
but I think we should do that anyway.

```suggestion
lemma maximalSeparatedSet_subset : maximalSeparatedSet ε A ⊆ A := by
  by grind [maximalSeparatedSet]
```